File: tests/csharp_static/TestClientCachePolicy.cs
using MyAirtable;
using Xunit;

namespace MyAirtable.Tests;

/// <summary>
/// CS11.1 — client read-cache policy (parity with the Java/Kotlin TestClientCachePolicy). Uses the
/// injectable <see cref="FakeTransport"/> to return canned 200s without the network. Like
/// Java/Kotlin, a list page carrying a live <c>offset</c> continuation token is NOT cached.
/// </summary>
public class TestClientCachePolicy
{
    private static AirtableClient ClientReturning(string body) =>
        new(
            "appX",
            "key",
            new FakeTransport((_, _) => FakeTransport.Canned.Ok(body)),
            cacheSeconds: 60.0
        );

    [Fact]
    public async Task SinglePagePayloadIsCached()
    {
        var client = ClientReturning("{\"records\": []}");
        await client.ListRecordsAsync("tbl1", new AirtableQuery());
        Assert.Equal(1, client.Cache.Count);
    }

    [Fact]
    public async Task GetRecordCachingIsUnaffected()
    {
        var client = ClientReturning("{\"id\": \"recX\", \"fields\": {}}");
        await client.GetRecordAsync("tbl1", "recX");
        Assert.Equal(1, client.Cache.Count);
    }

    [Fact]
    public async Task MultiPagePayloadWithOffsetIsNotCached()
    {
        // A page carrying a live `offset` continuation token must NOT be cached — the token
        // expires server-side, so a later hit would replay a dead token. Parity with Java/Kotlin.
        var client = ClientReturning("{\"records\": [], \"offset\": \"itrABC/recXYZ\"}");
        await client.ListRecordsAsync("tbl1", new AirtableQuery());
        Assert.Equal(0, client.Cache.Count);
    }

    [Fact]
    public async Task EmptyStringOffsetCountsAsComplete()
    {
        // An empty-string offset is not a continuation → the page is complete and cacheable.
        var client = ClientReturning("{\"records\": [], \"offset\": \"\"}");
        await client.ListRecordsAsync("tbl1", new AirtableQuery());
        Assert.Equal(1, client.Cache.Count);
    }

    [Fact]
    public async Task QueriesWithDelimitersInValuesDoNotCollideInCache()
    {
        // These two queries are structurally different but joined to the SAME cache key before the
        // fix (values were not encoded): filterByFormula="a&maxRecords=5" vs filterByFormula="a"
        // plus maxRecords=5. The collision served A's cached payload for B. Encoding both halves of
        // each pair makes the keys distinct.
        var n = 0;
        var client = new AirtableClient(
            "appX",
            "key",
            new FakeTransport(
                (_, _) => FakeTransport.Canned.Ok($"{{\"records\": [], \"n\": {n++}}}")
            ),
            cacheSeconds: 60.0
        );

        var a = await client.ListRecordsAsync(
            "tbl1",
            new AirtableQuery().WithFormula("a&maxRecords=5")
        );
        var b = await client.ListRecordsAsync(
            "tbl1",
            new AirtableQuery().WithFormula("a").WithMaxRecords(5)
        );

        Assert.NotEqual(a, b); // distinct payloads ⇒ B did not hit A's entry
        Assert.Equal(2, client.Cache.Count);
    }

    [Fact]
    public async Task IdenticalQueriesShareOneCacheEntry()
    {
        // A formula value containing `=` must still cache + hit correctly through the encoding.
        var calls = 0;
        var client = new AirtableClient(
            "appX",
            "key",
            new FakeTransport(
                (_, _) =>
                {
                    calls++;
                    return FakeTransport.Canned.Ok("{\"records\": []}");
                }
            ),
            cacheSeconds: 60.0
        );
        var query = new AirtableQuery().WithFormula("{f}=1").WithMaxRecords(5);

        await client.ListRecordsAsync("tbl1", query);
        await client.ListRecordsAsync("tbl1", query);

        Assert.Equal(1, calls); // second call served from cache
        Assert.Equal(1, client.Cache.Count);
    }
}

File: tests/cpp_static/test_client_cache_policy.cpp
// Client read-cache policy (Java TestClientCachePolicy parity + the C# offset-page skip).
#include <catch2/catch_test_macros.hpp>

#include "airtable_client.hpp"
#include "fake_transport.hpp"

using namespace myairtable;
using myairtable_tests::FakeTransport;

namespace {
AirtableClient cached_client(FakeTransport& transport, double cache_seconds = 60.0) {
    return AirtableClient("app1", "key1", transport.fn(), 0.0, 0.0, cache_seconds);
}
} // namespace

TEST_CASE("get_record is served from cache on the second call", "[client][cache]") {
    FakeTransport transport;
    transport.respond(200, R"({"id":"rec1"})");
    auto client = cached_client(transport);
    REQUIRE(client.get_record("tbl1", "rec1") == R"({"id":"rec1"})");
    REQUIRE(client.get_record("tbl1", "rec1") == R"({"id":"rec1"})");
    REQUIRE(transport.calls() == 1);
}

TEST_CASE("caching is off by default", "[client][cache]") {
    FakeTransport transport;
    transport.respond(200, "{}").respond(200, "{}");
    auto client = cached_client(transport, 0.0);
    client.get_record("tbl1", "rec1");
    client.get_record("tbl1", "rec1");
    REQUIRE(transport.calls() == 2);
}

TEST_CASE("list cache keys distinguish structurally different queries", "[client][cache]") {
    FakeTransport transport;
    transport.respond(200, R"({"records":[]})").respond(200, R"({"records":[]})");
    auto client = cached_client(transport);
    client.list_records("tbl1", AirtableQuery{.max_records = 1});
    client.list_records("tbl1", AirtableQuery{.max_records = 2});
    REQUIRE(transport.calls() == 2);
    client.list_records("tbl1", AirtableQuery{.max_records = 1}); // hit
    REQUIRE(transport.calls() == 2);
}

TEST_CASE("mutations invalidate the table's cached reads", "[client][cache]") {
    FakeTransport transport;
    transport.respond(200, R"({"id":"rec1","v":1})")
        .respond(200, R"({"records":[]})") // the create
        .respond(200, R"({"id":"rec1","v":2})");
    auto client = cached_client(transport);
    client.get_record("tbl1", "rec1");
    client.create_records("tbl1", R"({"records":[]})");
    REQUIRE(client.get_record("tbl1", "rec1") == R"({"id":"rec1","v":2})");
    REQUIRE(transport.calls() == 3);
}

TEST_CASE("a list page with a live offset token is never cached", "[client][cache]") {
    // The C# late-caught bug: caching a page whose offset token has expired
    // server-side would replay a dead token on the follow-up fetch.
    FakeTransport transport;
    transport.respond(200, R"({"records":[],"offset":"itrToken"})")
        .respond(200, R"({"records":[],"offset":"itrToken2"})");
    auto client = cached_client(transport);
    client.list_records("tbl1", AirtableQuery{});
    client.list_records("tbl1", AirtableQuery{});
    REQUIRE(transport.calls() == 2); // no cache hit
}

TEST_CASE("invalidate_all_caches clears every table", "[client][cache]") {
    FakeTransport transport;
    transport.respond(200, "{}").respond(200, "{}").respond(200, "{}").respond(200, "{}");
    auto client = cached_client(transport);
    client.get_record("tblA", "r1");
    client.get_record("tblB", "r2");
    client.invalidate_all_caches();
    client.get_record("tblA", "r1");
    client.get_record("tblB", "r2");
    REQUIRE(transport.calls() == 4);
}
